# PABPN1_phosphorylation

# Preprocessing, mapping, splicing, and poly(A) tail lengths quantification

Preprocessing, mapping, splicing, and poly(A) tail lengths quantification can be carried out with the attached Snakemake pipeline. Be sure to create/adjust the sample.yaml file as needed.

# Plotting and correlative analyses

Code for making plots and carrying out statistical and correlative comparisons (e.g. tail length vs. half-life) can be found the attached jupyter notebooks. 

Notebooks:
1. Alternative splicing 
2. Alternative polyadenylation
3. TimeLapse-seq/RNA stability
4. Differential gene expression (from DESeq2, Illumina sequencing)
5. Non-A nucleotide usage in poly(A) tails
6. Poly(A) tail lengths from all reads
7. Poly(A) tail lengths per gene
8. Poly(A) tail lengths per splice-isoform
9. Readlengths and coverage for LRS data
