{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis nonA nucleotide use in poly(A) tails"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n",
      "Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.\n"
     ]
    }
   ],
   "source": [
    "###Load Packages\n",
    "\n",
    "#Functions for polyA specific analysis (must be in same path)\n",
    "from polyA_utils import *\n",
    "\n",
    "#Packages for general column manipulation and statistics\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import csv\n",
    "from tqdm import tqdm\n",
    "from scipy import stats\n",
    "import scipy.stats as sci\n",
    "import pysam\n",
    "\n",
    "#Packages for plotting\n",
    "import matplotlib.pyplot as plt\n",
    "from statannotations.Annotator import Annotator\n",
    "import seaborn as sns\n",
    "sns.set(font_scale=1.5)\n",
    "sns.set_style(\"white\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Color palette for plots\n",
    "four_palette = [\"darkgreen\", \"tomato\", \"steelblue\", \"orange\"]\n",
    "two_palette = [\"steelblue\", \"orange\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "figpath = \"/Users/jacksongordon/Desktop/Papers/Yale_graduate/2023_PABPN1_manuscript/November_2023_edit/plots/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Question: Is there an increased propensity to use nonA nt's in any of the PABPN1 mutants?\n",
    "\n",
    "First read in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to read polyA tab del files\n",
    "def read_polyA_reads(file, min_read_count, sample_name, replicate):\n",
    "    df = pd.read_csv(file, delimiter='\\t')\n",
    "    df = df.loc[df['tail_len'] >= min_read_count]\n",
    "    df['Sample'] = sample_name\n",
    "    df['Replicate'] = replicate\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#path to the data\n",
    "path = \"/Users/jacksongordon/Desktop/Neugebauer_Lab/Experiments/Exp10_PABPN1_mutants_sequencing/Data_processing/Exp10/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read in dfs, assign to variables\n",
    "min_A_count = 10\n",
    "\n",
    "h112_1 = read_polyA_reads(path + 'results/polyA_len/h112.R1_polyA.txt', min_A_count, \"EV\", \"1\")\n",
    "h112_2 = read_polyA_reads(path + 'results/polyA_len/h112.R2_polyA.txt', min_A_count, \"EV\", \"2\")\n",
    "\n",
    "h113_1 = read_polyA_reads(path + 'results/polyA_len/h113.R1_polyA.txt', min_A_count, \"WT\", \"1\")\n",
    "h113_2 = read_polyA_reads(path + 'results/polyA_len/h113.R2_polyA.txt', min_A_count, \"WT\", \"2\")\n",
    "\n",
    "h115_1 = read_polyA_reads(path + 'results/polyA_len/h115.R1_polyA.txt', min_A_count, \"4SA\", \"1\")\n",
    "h115_2 = read_polyA_reads(path + 'results/polyA_len/h115.R2_polyA.txt', min_A_count, \"4SA\", \"2\")\n",
    "\n",
    "h117_1 = read_polyA_reads(path + 'results/polyA_len/h117.R1_polyA.txt', min_A_count, \"4SD\", \"1\")\n",
    "h117_2 = read_polyA_reads(path + 'results/polyA_len/h117.R2_polyA.txt', min_A_count, \"4SD\", \"2\")\n",
    "\n",
    "#Path to gene types df (from gtf)\n",
    "gene_types = pd.read_csv(path + \"config/gene_types_df.tsv\", sep = \"\\t\")\n",
    "\n",
    "#Combine replicates and intersect with gene names\n",
    "all_reads = pd.concat([h112_1, h112_2, h113_1, h113_2, h115_1, h115_2, h117_1, h117_2], ignore_index = True)\n",
    "all_reads[\"gene_id\"] = all_reads[\"gene_id\"].str.split(\".\").str[0]\n",
    "all_reads = all_reads.merge(gene_types, on = 'gene_id')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine whether there is nonA usage in tails. First align the first or last 400 nt of poly(A) tails and calculate the percent of each nucleotide usage for each condition. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                                                                              | 0/718631 [00:00<?, ?it/s]Fatal Python error: config_get_locale_encoding: failed to get the locale encoding: nl_langinfo(CODESET) failed\n",
      "Python runtime state: preinitialized\n",
      "\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 718631/718631 [04:06<00:00, 2912.41it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 496934/496934 [02:55<00:00, 2836.10it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 532471/532471 [03:29<00:00, 2547.19it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 141151/141151 [00:45<00:00, 3126.12it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 334477/334477 [02:06<00:00, 2644.41it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 167835/167835 [01:05<00:00, 2564.43it/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 268798/268798 [22:06<00:00, 202.57it/s]\n",
      " 63%|████████████████████████████████████████████████████████████████████████████████▊                                               | 76468/121037 [00:30<00:14, 3026.96it/s]"
     ]
    }
   ],
   "source": [
    "#Non A bases - run this cell once \n",
    "#trange defines the number of bases to consider in the tail alignment\n",
    "trange = 400\n",
    "from polyA_utils import *\n",
    "\n",
    "total_c_h112_1, tail_len_h112_1, base_count_prox_h112_1, base_count_dist_h112_1, base_cov_h112_1 = nonA(path + 'results/mapping/h112.R1_hg38.bam', tail_range = trange)\n",
    "total_c_h113_1, tail_len_h113_1, base_count_prox_h113_1, base_count_dist_h113_1, base_cov_h113_1 = nonA(path + 'results/mapping/h113.R1_hg38.bam', tail_range = trange)\n",
    "total_c_h115_1, tail_len_h115_1, base_count_prox_h115_1, base_count_dist_h115_1, base_cov_h115_1 = nonA(path + 'results/mapping/h115.R1_hg38.bam', tail_range = trange)\n",
    "total_c_h117_1, tail_len_h117_1, base_count_prox_h117_1, base_count_dist_h117_1, base_cov_h117_1 = nonA(path + 'results/mapping/h117.R1_hg38.bam', tail_range = trange)\n",
    "\n",
    "total_c_h112_2, tail_len_h112_2, base_count_prox_h112_2, base_count_dist_h112_2, base_cov_h112_2 = nonA(path + 'results/mapping/h112.R2_hg38.bam', tail_range = trange)\n",
    "total_c_h113_2, tail_len_h113_2, base_count_prox_h113_2, base_count_dist_h113_2, base_cov_h113_2 = nonA(path + 'results/mapping/h113.R2_hg38.bam', tail_range = trange)\n",
    "total_c_h115_2, tail_len_h115_2, base_count_prox_h115_2, base_count_dist_h115_2, base_cov_h115_2 = nonA(path + 'results/mapping/h115.R2_hg38.bam', tail_range = trange)\n",
    "total_c_h117_2, tail_len_h117_2, base_count_prox_h117_2, base_count_dist_h117_2, base_cov_h117_2 = nonA(path + 'results/mapping/h117.R2_hg38.bam', tail_range = trange)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Are there any differences in the percent of each base used between conditions?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nt_df = pd.concat([\n",
    "    pd.DataFrame.from_dict(total_c_h112_1, orient = \"index\").transpose(),\n",
    "    pd.DataFrame.from_dict(total_c_h112_2, orient = \"index\").transpose(),\n",
    "\n",
    "    pd.DataFrame.from_dict(total_c_h113_1, orient = \"index\").transpose(),\n",
    "    pd.DataFrame.from_dict(total_c_h113_2, orient = \"index\").transpose(),\n",
    "\n",
    "    pd.DataFrame.from_dict(total_c_h115_1, orient = \"index\").transpose(),\n",
    "    pd.DataFrame.from_dict(total_c_h115_2, orient = \"index\").transpose(),\n",
    "\n",
    "    pd.DataFrame.from_dict(total_c_h117_1, orient = \"index\").transpose(),\n",
    "    pd.DataFrame.from_dict(total_c_h117_2, orient = \"index\").transpose()\n",
    "        ])\n",
    "\n",
    "nt_df = nt_df.div(nt_df.sum(axis=1), axis=0)\n",
    "\n",
    "nt_df[\"Sample\"] = [\"EV\", \"EV\", \"WT\", \"WT\", \"4SA\", \"4SA\", \"4SD\", \"4SD\"]\n",
    "nt_df[\"Replicate\"] = [1, 2, 1, 2, 1, 2, 1, 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, ax = plt.subplots(1,1, figsize = (5,5))\n",
    "\n",
    "sns.stripplot(x = nt_df[\"Sample\"], y = nt_df[\"C\"], color = \"darkgreen\", size = 7)\n",
    "sns.stripplot(x = nt_df[\"Sample\"], y = nt_df[\"G\"], color = \"steelblue\", size = 7)\n",
    "sns.stripplot(x = nt_df[\"Sample\"], y = nt_df[\"T\"], color = \"tomato\", size = 7)\n",
    "\n",
    "ax.set_ylabel(\"percent nt usage\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First create a dataframe of the nonA nt usasge calculated above for ease of plot generation. Then generate plots to look at the percent of each nucleotide used across the alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to create a dataframe from tail composition data\n",
    "def nonA_df(total_c, tail_len, base_count_prox, base_count_dist, base_cov):\n",
    "    #EV - 5' alignment\n",
    "    x=np.arange(1, 401)\n",
    "    percent_5 = pd.DataFrame()\n",
    "    percent_5['A'] = base_count_prox[0]/base_cov\n",
    "    percent_5['G'] = base_count_prox[1]/base_cov\n",
    "    percent_5['U'] = base_count_prox[2]/base_cov\n",
    "    percent_5['C'] = base_count_prox[3]/base_cov\n",
    "    percent_5['reads'] = base_cov\n",
    "    #EV - 3' alignment\n",
    "    x=np.arange(1, 401)\n",
    "    percent_3 = pd.DataFrame()\n",
    "    percent_3['A'] = base_count_dist[0]/base_cov[::-1]\n",
    "    percent_3['G'] = base_count_dist[1]/base_cov[::-1]\n",
    "    percent_3['U'] = base_count_dist[2]/base_cov[::-1]\n",
    "    percent_3['C'] = base_count_dist[3]/base_cov[::-1]\n",
    "    percent_3['reads'] = base_cov[::-1]\n",
    "    return(percent_5, percent_3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the function on all conditions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#EV\n",
    "h112_1_df_5, h112_1_df_3 = nonA_df(total_c_h112_1, tail_len_h112_1, base_count_prox_h112_1, base_count_dist_h112_1, base_cov_h112_1)\n",
    "h112_2_df_5, h112_2_df_3 = nonA_df(total_c_h112_2, tail_len_h112_2, base_count_prox_h112_2, base_count_dist_h112_2, base_cov_h112_2)\n",
    "tmp = pd.concat((h112_1_df_5, h112_2_df_5))\n",
    "h112_percent_5 = tmp.groupby(tmp.index).mean()\n",
    "tmp = pd.concat((h112_1_df_3, h112_2_df_3))\n",
    "h112_percent_3 = tmp.groupby(tmp.index).mean()\n",
    "#WT\n",
    "h113_1_df_5, h113_1_df_3 = nonA_df(total_c_h113_1, tail_len_h113_1, base_count_prox_h113_1, base_count_dist_h113_1, base_cov_h113_1)\n",
    "h113_2_df_5, h113_2_df_3 = nonA_df(total_c_h113_2, tail_len_h113_2, base_count_prox_h113_2, base_count_dist_h113_2, base_cov_h113_2)\n",
    "tmp = pd.concat((h113_1_df_5, h113_2_df_5))\n",
    "h113_percent_5 = tmp.groupby(tmp.index).mean()\n",
    "tmp = pd.concat((h113_1_df_3, h113_2_df_3))\n",
    "h113_percent_3 = tmp.groupby(tmp.index).mean()\n",
    "#4SA\n",
    "h115_1_df_5, h115_1_df_3 = nonA_df(total_c_h115_1, tail_len_h115_1, base_count_prox_h115_1, base_count_dist_h115_1, base_cov_h115_1)\n",
    "h115_2_df_5, h115_2_df_3 = nonA_df(total_c_h115_2, tail_len_h115_2, base_count_prox_h115_2, base_count_dist_h115_2, base_cov_h115_2)\n",
    "tmp = pd.concat((h115_1_df_5, h115_2_df_5))\n",
    "h115_percent_5 = tmp.groupby(tmp.index).mean()\n",
    "tmp = pd.concat((h115_1_df_3, h115_2_df_3))\n",
    "h115_percent_3 = tmp.groupby(tmp.index).mean()\n",
    "#4SD\n",
    "h117_1_df_5, h117_1_df_3 = nonA_df(total_c_h117_1, tail_len_h117_1, base_count_prox_h117_1, base_count_dist_h117_1, base_cov_h117_1)\n",
    "h117_2_df_5, h117_2_df_3 = nonA_df(total_c_h117_2, tail_len_h117_2, base_count_prox_h117_2, base_count_dist_h117_2, base_cov_h117_2)\n",
    "tmp = pd.concat((h117_1_df_5, h117_2_df_5))\n",
    "h117_percent_5 = tmp.groupby(tmp.index).mean()\n",
    "tmp = pd.concat((h117_1_df_3, h117_2_df_3))\n",
    "h117_percent_3 = tmp.groupby(tmp.index).mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now plot the percent use of each nucleotide for each condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Plot EV data - all reads\n",
    "x = np.arange(1, 401)\n",
    "\n",
    "fig, (ax1, ax3) = plt.subplots(1,2, figsize=(10,5))\n",
    "\n",
    "ax1 = sns.scatterplot(data = h112_percent_5, x=x, y='G', alpha=.5, s = 50, color = 'darkgreen', ax = ax1)\n",
    "ax1 = sns.scatterplot(data = h112_percent_5, x=x, y='U', alpha=.5, s = 50, color = 'tomato', ax = ax1)\n",
    "ax1 = sns.scatterplot(data = h112_percent_5, x=x, y='C', alpha=.5, s = 50, color = 'steelblue', ax = ax1)\n",
    "ax1.set_xlim([-1, 400])\n",
    "ax1.set_ylim([0, 0.003])\n",
    "#ax1.set_yscale('log')\n",
    "ax1.set_ylabel('fraction')\n",
    "ax1.legend(['G', 'U', 'C'])\n",
    "ax1.set_title(\"5' aligned\")\n",
    "ax3.set_title(\"3' aligned\")\n",
    "#ax.set_title(\"S->D 5' tail\")\n",
    "ax2 = ax1.twinx()\n",
    "ax2 = sns.lineplot(data = h112_percent_5, x=x, y='reads', alpha=.5, color = 'grey', lw = 3)\n",
    "\n",
    "ax3 = sns.scatterplot(data = h112_percent_3, x=x, y='G', alpha=.5, s = 50, color = 'darkgreen', ax = ax3)\n",
    "ax3 = sns.scatterplot(data = h112_percent_3, x=x, y='U', alpha=.5, s = 50, color = 'tomato', ax = ax3)\n",
    "ax3 = sns.scatterplot(data = h112_percent_3, x=x, y='C', alpha=.5, s = 50, color = 'steelblue', ax = ax3)\n",
    "ax3.set_xlim([0, 400])\n",
    "ax3.set_ylim([0, 0.003])\n",
    "#ax1.set_yscale('log')\n",
    "ax3.set_ylabel('fraction')\n",
    "ax3.legend(['G', 'U', 'C'])\n",
    "#ax.set_title(\"S->D 5' tail\")\n",
    "ax4 = ax3.twinx()\n",
    "ax4 = sns.lineplot(data = h112_percent_3, x=x, y='reads', alpha=.5, color = 'grey', lw = 3)\n",
    "ax1.set_xlabel(\"nucleotide position\")\n",
    "ax3.set_xlabel(\"nucleotide position\")\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot WT data - all reads\n",
    "x = np.arange(1, 401)\n",
    "\n",
    "fig, (ax1, ax3) = plt.subplots(1,2, figsize=(10,5))\n",
    "\n",
    "ax1 = sns.scatterplot(data = h113_percent_5, x=x, y='G', alpha=.5, s = 50, color = 'darkgreen', ax = ax1)\n",
    "ax1 = sns.scatterplot(data = h113_percent_5, x=x, y='U', alpha=.5, s = 50, color = 'tomato', ax = ax1)\n",
    "ax1 = sns.scatterplot(data = h113_percent_5, x=x, y='C', alpha=.5, s = 50, color = 'steelblue', ax = ax1)\n",
    "ax1.set_xlim([-1, 400])\n",
    "ax1.set_ylim([0, 0.003])\n",
    "#ax1.set_yscale('log')\n",
    "ax1.set_ylabel('fraction')\n",
    "ax1.legend(['G', 'U', 'C'])\n",
    "ax1.set_title(\"5' aligned\")\n",
    "ax3.set_title(\"3' aligned\")\n",
    "#ax.set_title(\"S->D 5' tail\")\n",
    "ax2 = ax1.twinx()\n",
    "ax2 = sns.lineplot(data = h113_percent_5, x=x, y='reads', alpha=.5, color = 'grey', lw = 3)\n",
    "\n",
    "ax3 = sns.scatterplot(data = h113_percent_3, x=x, y='G', alpha=.5, s = 50, color = 'darkgreen', ax = ax3)\n",
    "ax3 = sns.scatterplot(data = h113_percent_3, x=x, y='U', alpha=.5, s = 50, color = 'tomato', ax = ax3)\n",
    "ax3 = sns.scatterplot(data = h113_percent_3, x=x, y='C', alpha=.5, s = 50, color = 'steelblue', ax = ax3)\n",
    "ax3.set_xlim([0, 400])\n",
    "ax3.set_ylim([0, 0.003])\n",
    "#ax1.set_yscale('log')\n",
    "ax3.set_ylabel('fraction')\n",
    "ax3.legend(['G', 'U', 'C'])\n",
    "#ax.set_title(\"S->D 5' tail\")\n",
    "ax4 = ax3.twinx()\n",
    "ax4 = sns.lineplot(data = h113_percent_3, x=x, y='reads', alpha=.5, color = 'grey', lw = 3)\n",
    "ax1.set_xlabel(\"nucleotide position\")\n",
    "ax3.set_xlabel(\"nucleotide position\")\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot 4SA data - all reads\n",
    "x = np.arange(1, 401)\n",
    "\n",
    "fig, (ax1, ax3) = plt.subplots(1,2, figsize=(10,5))\n",
    "\n",
    "ax1 = sns.scatterplot(data = h115_percent_5, x=x, y='G', alpha=.5, s = 50, color = 'darkgreen', ax = ax1)\n",
    "ax1 = sns.scatterplot(data = h115_percent_5, x=x, y='U', alpha=.5, s = 50, color = 'tomato', ax = ax1)\n",
    "ax1 = sns.scatterplot(data = h115_percent_5, x=x, y='C', alpha=.5, s = 50, color = 'steelblue', ax = ax1)\n",
    "ax1.set_xlim([-1, 400])\n",
    "ax1.set_ylim([0, 0.003])\n",
    "#ax1.set_yscale('log')\n",
    "ax1.set_ylabel('fraction')\n",
    "ax1.legend(['G', 'U', 'C'])\n",
    "ax1.set_title(\"5' aligned\")\n",
    "ax3.set_title(\"3' aligned\")\n",
    "#ax.set_title(\"S->D 5' tail\")\n",
    "ax2 = ax1.twinx()\n",
    "ax2 = sns.lineplot(data = h115_percent_5, x=x, y='reads', alpha=.5, color = 'grey', lw = 3)\n",
    "\n",
    "ax3 = sns.scatterplot(data = h115_percent_3, x=x, y='G', alpha=.5, s = 50, color = 'darkgreen', ax = ax3)\n",
    "ax3 = sns.scatterplot(data = h115_percent_3, x=x, y='U', alpha=.5, s = 50, color = 'tomato', ax = ax3)\n",
    "ax3 = sns.scatterplot(data = h115_percent_3, x=x, y='C', alpha=.5, s = 50, color = 'steelblue', ax = ax3)\n",
    "ax3.set_xlim([0, 400])\n",
    "ax3.set_ylim([0, 0.003])\n",
    "#ax1.set_yscale('log')\n",
    "ax3.set_ylabel('fraction')\n",
    "ax3.legend(['G', 'U', 'C'])\n",
    "#ax.set_title(\"S->D 5' tail\")\n",
    "ax4 = ax3.twinx()\n",
    "ax4 = sns.lineplot(data = h115_percent_3, x=x, y='reads', alpha=.5, color = 'grey', lw = 3)\n",
    "ax1.set_xlabel(\"nucleotide position\")\n",
    "ax3.set_xlabel(\"nucleotide position\")\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot 4SD data - all reads\n",
    "x = np.arange(1, 401)\n",
    "\n",
    "fig, (ax1, ax3) = plt.subplots(1,2, figsize=(10,5))\n",
    "\n",
    "ax1 = sns.scatterplot(data = h117_percent_5, x=x, y='G', alpha=.5, s = 50, color = 'darkgreen', ax = ax1)\n",
    "ax1 = sns.scatterplot(data = h117_percent_5, x=x, y='U', alpha=.5, s = 50, color = 'tomato', ax = ax1)\n",
    "ax1 = sns.scatterplot(data = h117_percent_5, x=x, y='C', alpha=.5, s = 50, color = 'steelblue', ax = ax1)\n",
    "ax1.set_xlim([-1, 400])\n",
    "ax1.set_ylim([0, 0.003])\n",
    "#ax1.set_yscale('log')\n",
    "ax1.set_ylabel('fraction')\n",
    "ax1.legend(['G', 'U', 'C'])\n",
    "ax1.set_title(\"5' aligned\")\n",
    "ax3.set_title(\"3' aligned\")\n",
    "#ax.set_title(\"S->D 5' tail\")\n",
    "ax2 = ax1.twinx()\n",
    "ax2 = sns.lineplot(data = h117_percent_5, x=x, y='reads', alpha=.5, color = 'grey', lw = 3)\n",
    "\n",
    "ax3 = sns.scatterplot(data = h117_percent_3, x=x, y='G', alpha=.5, s = 50, color = 'darkgreen', ax = ax3)\n",
    "ax3 = sns.scatterplot(data = h117_percent_3, x=x, y='U', alpha=.5, s = 50, color = 'tomato', ax = ax3)\n",
    "ax3 = sns.scatterplot(data = h117_percent_3, x=x, y='C', alpha=.5, s = 50, color = 'steelblue', ax = ax3)\n",
    "ax3.set_xlim([0, 400])\n",
    "ax3.set_ylim([0, 0.003])\n",
    "#ax1.set_yscale('log')\n",
    "ax3.set_ylabel('fraction')\n",
    "ax3.legend(['G', 'U', 'C'])\n",
    "#ax.set_title(\"S->D 5' tail\")\n",
    "ax4 = ax3.twinx()\n",
    "ax4 = sns.lineplot(data = h117_percent_3, x=x, y='reads', alpha=.5, color = 'grey', lw = 3)\n",
    "ax1.set_xlabel(\"nucleotide position\")\n",
    "ax3.set_xlabel(\"nucleotide position\")\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The alignments will be noisly looking at position-to-position differences. Define a function to create bins in which to look at differences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bin_array(mut, cov, binsize, **kwargs):\n",
    "    # reactivity calculation in discrete bins of polymerase positions\n",
    "    assert len(mut) == len(cov)\n",
    "\n",
    "    bins = np.arange(0, len(mut), binsize)\n",
    "    mut_bin = np.zeros(len(bins))\n",
    "    cov_bin = np.zeros(len(bins))\n",
    "\n",
    "    for i in range(len(bins)):\n",
    "        if i < len(bins)-1:\n",
    "            cov_bin[i] = np.sum(cov[bins[i]:bins[i+1]])\n",
    "            mut_bin[i] = np.sum(mut[bins[i]:bins[i+1]])\n",
    "\n",
    "        else: # last bin in case it doesn't add up\n",
    "            cov_bin[i] = np.sum(cov[bins[i]:])\n",
    "            mut_bin[i] = np.sum(mut[bins[i]:])\n",
    "            \n",
    "    return(bins, mut_bin, cov_bin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#EV\n",
    "bins = 5\n",
    "base_count_prox_a_h112_1_bins = bin_array(base_count_prox_h112_1[0], base_cov_h112_1, bins)\n",
    "base_count_prox_a_h112_2_bins = bin_array(base_count_prox_h112_2[0], base_cov_h112_2, bins)\n",
    "\n",
    "base_count_prox_g_h112_1_bins = bin_array(base_count_prox_h112_1[1], base_cov_h112_1, bins)\n",
    "base_count_prox_g_h112_2_bins = bin_array(base_count_prox_h112_2[1], base_cov_h112_2, bins)\n",
    "\n",
    "base_count_prox_u_h112_1_bins = bin_array(base_count_prox_h112_1[2], base_cov_h112_1, bins)\n",
    "base_count_prox_u_h112_2_bins = bin_array(base_count_prox_h112_2[2], base_cov_h112_2, bins)\n",
    "\n",
    "base_count_prox_c_h112_1_bins = bin_array(base_count_prox_h112_1[3], base_cov_h112_1, bins)\n",
    "base_count_prox_c_h112_2_bins = bin_array(base_count_prox_h112_2[3], base_cov_h112_2, bins)\n",
    "\n",
    "base_count_dist_a_h112_1_bins = bin_array(base_count_dist_h112_1[0], base_cov_h112_1, bins)\n",
    "base_count_dist_a_h112_2_bins = bin_array(base_count_dist_h112_2[0], base_cov_h112_2, bins)\n",
    "\n",
    "base_count_dist_g_h112_1_bins = bin_array(base_count_dist_h112_1[1], base_cov_h112_1, bins)\n",
    "base_count_dist_g_h112_2_bins = bin_array(base_count_dist_h112_2[1], base_cov_h112_2, bins)\n",
    "\n",
    "base_count_dist_u_h112_1_bins = bin_array(base_count_dist_h112_1[2], base_cov_h112_1, bins)\n",
    "base_count_dist_u_h112_2_bins = bin_array(base_count_dist_h112_2[2], base_cov_h112_2, bins)\n",
    "\n",
    "base_count_dist_c_h112_1_bins = bin_array(base_count_dist_h112_1[3], base_cov_h112_1, bins)\n",
    "base_count_dist_c_h112_2_bins = bin_array(base_count_dist_h112_2[3], base_cov_h112_2, bins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#WT\n",
    "bins = 5\n",
    "base_count_prox_a_h113_1_bins = bin_array(base_count_prox_h113_1[0], base_cov_h113_1, bins)\n",
    "base_count_prox_a_h113_2_bins = bin_array(base_count_prox_h113_2[0], base_cov_h113_2, bins)\n",
    "\n",
    "base_count_prox_g_h113_1_bins = bin_array(base_count_prox_h113_1[1], base_cov_h113_1, bins)\n",
    "base_count_prox_g_h113_2_bins = bin_array(base_count_prox_h113_2[1], base_cov_h113_2, bins)\n",
    "\n",
    "base_count_prox_u_h113_1_bins = bin_array(base_count_prox_h113_1[2], base_cov_h113_1, bins)\n",
    "base_count_prox_u_h113_2_bins = bin_array(base_count_prox_h113_2[2], base_cov_h113_2, bins)\n",
    "\n",
    "base_count_prox_c_h113_1_bins = bin_array(base_count_prox_h113_1[3], base_cov_h113_1, bins)\n",
    "base_count_prox_c_h113_2_bins = bin_array(base_count_prox_h113_2[3], base_cov_h113_2, bins)\n",
    "\n",
    "base_count_dist_a_h113_1_bins = bin_array(base_count_dist_h113_1[0], base_cov_h113_1, bins)\n",
    "base_count_dist_a_h113_2_bins = bin_array(base_count_dist_h113_2[0], base_cov_h113_2, bins)\n",
    "\n",
    "base_count_dist_g_h113_1_bins = bin_array(base_count_dist_h113_1[1], base_cov_h113_1, bins)\n",
    "base_count_dist_g_h113_2_bins = bin_array(base_count_dist_h113_2[1], base_cov_h113_2, bins)\n",
    "\n",
    "base_count_dist_u_h113_1_bins = bin_array(base_count_dist_h113_1[2], base_cov_h113_1, bins)\n",
    "base_count_dist_u_h113_2_bins = bin_array(base_count_dist_h113_2[2], base_cov_h113_2, bins)\n",
    "\n",
    "base_count_dist_c_h113_1_bins = bin_array(base_count_dist_h113_1[3], base_cov_h113_1, bins)\n",
    "base_count_dist_c_h113_2_bins = bin_array(base_count_dist_h113_2[3], base_cov_h113_2, bins)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4SA\n",
    "bins = 5\n",
    "base_count_prox_a_h115_1_bins = bin_array(base_count_prox_h115_1[0], base_cov_h115_1, bins)\n",
    "base_count_prox_a_h115_2_bins = bin_array(base_count_prox_h115_2[0], base_cov_h115_2, bins)\n",
    "\n",
    "base_count_prox_g_h115_1_bins = bin_array(base_count_prox_h115_1[1], base_cov_h115_1, bins)\n",
    "base_count_prox_g_h115_2_bins = bin_array(base_count_prox_h115_2[1], base_cov_h115_2, bins)\n",
    "\n",
    "base_count_prox_u_h115_1_bins = bin_array(base_count_prox_h115_1[2], base_cov_h115_1, bins)\n",
    "base_count_prox_u_h115_2_bins = bin_array(base_count_prox_h115_2[2], base_cov_h115_2, bins)\n",
    "\n",
    "base_count_prox_c_h115_1_bins = bin_array(base_count_prox_h115_1[3], base_cov_h115_1, bins)\n",
    "base_count_prox_c_h115_2_bins = bin_array(base_count_prox_h115_2[3], base_cov_h115_2, bins)\n",
    "\n",
    "base_count_dist_a_h115_1_bins = bin_array(base_count_dist_h115_1[0], base_cov_h115_1, bins)\n",
    "base_count_dist_a_h115_2_bins = bin_array(base_count_dist_h115_2[0], base_cov_h115_2, bins)\n",
    "\n",
    "base_count_dist_g_h115_1_bins = bin_array(base_count_dist_h115_1[1], base_cov_h115_1, bins)\n",
    "base_count_dist_g_h115_2_bins = bin_array(base_count_dist_h115_2[1], base_cov_h115_2, bins)\n",
    "\n",
    "base_count_dist_u_h115_1_bins = bin_array(base_count_dist_h115_1[2], base_cov_h115_1, bins)\n",
    "base_count_dist_u_h115_2_bins = bin_array(base_count_dist_h115_2[2], base_cov_h115_2, bins)\n",
    "\n",
    "base_count_dist_c_h115_1_bins = bin_array(base_count_dist_h115_1[3], base_cov_h115_1, bins)\n",
    "base_count_dist_c_h115_2_bins = bin_array(base_count_dist_h115_2[3], base_cov_h115_2, bins)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4SD\n",
    "bins = 5\n",
    "base_count_prox_a_h117_1_bins = bin_array(base_count_prox_h117_1[0], base_cov_h117_1, bins)\n",
    "base_count_prox_a_h117_2_bins = bin_array(base_count_prox_h117_2[0], base_cov_h117_2, bins)\n",
    "\n",
    "base_count_prox_g_h117_1_bins = bin_array(base_count_prox_h117_1[1], base_cov_h117_1, bins)\n",
    "base_count_prox_g_h117_2_bins = bin_array(base_count_prox_h117_2[1], base_cov_h117_2, bins)\n",
    "\n",
    "base_count_prox_u_h117_1_bins = bin_array(base_count_prox_h117_1[2], base_cov_h117_1, bins)\n",
    "base_count_prox_u_h117_2_bins = bin_array(base_count_prox_h117_2[2], base_cov_h117_2, bins)\n",
    "\n",
    "base_count_prox_c_h117_1_bins = bin_array(base_count_prox_h117_1[3], base_cov_h117_1, bins)\n",
    "base_count_prox_c_h117_2_bins = bin_array(base_count_prox_h117_2[3], base_cov_h117_2, bins)\n",
    "\n",
    "base_count_dist_a_h117_1_bins = bin_array(base_count_dist_h117_1[0], base_cov_h117_1, bins)\n",
    "base_count_dist_a_h117_2_bins = bin_array(base_count_dist_h117_2[0], base_cov_h117_2, bins)\n",
    "\n",
    "base_count_dist_g_h117_1_bins = bin_array(base_count_dist_h117_1[1], base_cov_h117_1, bins)\n",
    "base_count_dist_g_h117_2_bins = bin_array(base_count_dist_h117_2[1], base_cov_h117_2, bins)\n",
    "\n",
    "base_count_dist_u_h117_1_bins = bin_array(base_count_dist_h117_1[2], base_cov_h117_1, bins)\n",
    "base_count_dist_u_h117_2_bins = bin_array(base_count_dist_h117_2[2], base_cov_h117_2, bins)\n",
    "\n",
    "base_count_dist_c_h117_1_bins = bin_array(base_count_dist_h117_1[3], base_cov_h117_1, bins)\n",
    "base_count_dist_c_h117_2_bins = bin_array(base_count_dist_h117_2[3], base_cov_h117_2, bins)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(base_count_dist_u_h115_1_bins[1]/(base_count_dist_u_h115_1_bins[2] + base_count_dist_u_h115_1_bins[1]) + \n",
    "base_count_dist_u_h115_2_bins[1]/(base_count_dist_u_h115_2_bins[2] + base_count_dist_u_h115_2_bins[1]))/2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generate figures to compare nonA use between two conditions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "C usage - stats comparing 4SA to 4SD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run t-test between percent nt ussage at each postion\n",
    "stat_vals_5 = stats.ttest_ind(\n",
    "                            [base_count_dist_c_h115_1_bins[1]/(base_count_dist_c_h115_1_bins[2] + base_count_dist_c_h115_1_bins[1]), base_count_dist_c_h115_2_bins[1]/(base_count_dist_c_h115_2_bins[2] + base_count_dist_c_h115_2_bins[1])],\n",
    "                            [base_count_dist_c_h117_1_bins[1]/(base_count_dist_c_h117_1_bins[2] + base_count_dist_c_h117_1_bins[1]), base_count_dist_c_h117_2_bins[1]/(base_count_dist_c_h117_2_bins[2] + (base_count_dist_c_h117_2_bins[1]))])[-1]\n",
    "\n",
    "#calculate FRD\n",
    "stat_vals_corr_5 = stats.false_discovery_control(stat_vals_5)\n",
    "\n",
    "#Run t-test between percent nt ussage at each postion\n",
    "stat_vals_3 = stats.ttest_ind(\n",
    "                            [base_count_prox_c_h115_1_bins[1]/(base_count_prox_c_h115_1_bins[2] + base_count_prox_c_h115_1_bins[1]), base_count_prox_c_h115_2_bins[1]/(base_count_prox_c_h115_2_bins[2] + base_count_prox_c_h115_2_bins[1])],\n",
    "                            [base_count_prox_c_h117_1_bins[1]/(base_count_prox_c_h117_1_bins[2] + base_count_prox_c_h117_1_bins[1]), base_count_prox_c_h117_2_bins[1]/(base_count_prox_c_h117_2_bins[2] + (base_count_prox_c_h117_2_bins[1]))])[-1]\n",
    "\n",
    "#calculate FRD\n",
    "stat_vals_corr_3 = stats.false_discovery_control(stat_vals_3)\n",
    "\n",
    "#############################################################\n",
    "\n",
    "#Figure for percent nt - 5' end A\n",
    "#Generate plot\n",
    "fig, (ax2, axs) = plt.subplots(2, 2, gridspec_kw={'height_ratios': [1, 3]}, figsize=(10, 5))\n",
    "\n",
    "#distal\n",
    "sns.lineplot(x = (base_count_dist_c_h112_1_bins[0] + base_count_dist_c_h112_2_bins[0])/2, \n",
    "             y=(base_count_dist_c_h112_1_bins[1]/(base_count_dist_c_h112_1_bins[2] + base_count_dist_c_h112_1_bins[1])) + \n",
    "               (base_count_dist_c_h112_2_bins[1]/(base_count_dist_c_h112_2_bins[2] + base_count_dist_c_h112_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'darkgreen', ax = axs[1], linewidth = 5)\n",
    "\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_c_h113_1_bins[0] + base_count_dist_c_h113_2_bins[0])/2, \n",
    "             y=(base_count_dist_c_h113_1_bins[1]/(base_count_dist_c_h113_1_bins[2] + base_count_dist_c_h113_1_bins[1])) + \n",
    "               (base_count_dist_c_h113_2_bins[1]/(base_count_dist_c_h113_2_bins[2] + base_count_dist_c_h113_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'tomato', ax = axs[1], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_c_h115_1_bins[0] + base_count_dist_c_h115_2_bins[0])/2, \n",
    "             y=(base_count_dist_c_h115_1_bins[1]/(base_count_dist_c_h115_1_bins[2] + base_count_dist_c_h115_1_bins[1])) + \n",
    "               (base_count_dist_c_h115_2_bins[1]/(base_count_dist_c_h115_2_bins[2] + base_count_dist_c_h115_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'steelblue', ax = axs[1], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_c_h117_1_bins[0] + base_count_dist_c_h117_2_bins[0])/2, \n",
    "             y=(base_count_dist_c_h117_1_bins[1]/(base_count_dist_c_h117_1_bins[2] + base_count_dist_c_h117_1_bins[1])) + \n",
    "               (base_count_dist_c_h117_2_bins[1]/(base_count_dist_c_h117_2_bins[2] + base_count_dist_c_h117_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'orange', ax = axs[1], linewidth = 5)\n",
    "\n",
    "\n",
    "ax2[1] = sns.lineplot(x = (base_count_dist_c_h115_1_bins[0] + base_count_dist_c_h115_2_bins[0])/2, \n",
    "                   y = -np.log10(stat_vals_5), \n",
    "                   alpha=.8, color = 'tomato', ax = ax2[1], linewidth = 5)\n",
    "\n",
    "ax2[1].set_ylabel(\"-log10(p)\")\n",
    "ax2[1].axhline(-np.log10(0.05), linestyle = '--', color = 'black')\n",
    "\n",
    "axs[1].set_xlim(10, 390)\n",
    "ax2[1].set_xlim(10, 390)\n",
    "ax2[1].set(xticklabels=[])  # remove the tick labels\n",
    "ax2[1].tick_params(bottom=False)  # remove the ticks\n",
    "\n",
    "axs[1].set_ylim(0, 0.1)\n",
    "ax2[1].set_ylim(0,3)\n",
    "\n",
    "axs[1].set_ylabel(\"percent\")\n",
    "axs[1].set_xlabel(\"distance from 3' end (nt)\")\n",
    "\n",
    "###############################################\n",
    "#proximal\n",
    "sns.lineplot(x = (base_count_prox_c_h112_1_bins[0] + base_count_prox_c_h112_2_bins[0])/2, \n",
    "             y=(base_count_prox_c_h112_1_bins[1]/(base_count_prox_c_h112_1_bins[2] + base_count_prox_c_h112_1_bins[1])) + \n",
    "               (base_count_prox_c_h112_2_bins[1]/(base_count_prox_c_h112_2_bins[2] + base_count_prox_c_h112_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'darkgreen', ax = axs[0], linewidth = 5)\n",
    "\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_c_h113_1_bins[0] + base_count_prox_c_h113_2_bins[0])/2, \n",
    "             y=(base_count_prox_c_h113_1_bins[1]/(base_count_prox_c_h113_1_bins[2] + base_count_prox_c_h113_1_bins[1])) + \n",
    "               (base_count_prox_c_h113_2_bins[1]/(base_count_prox_c_h113_2_bins[2] + base_count_prox_c_h113_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'tomato', ax = axs[0], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_c_h115_1_bins[0] + base_count_prox_c_h115_2_bins[0])/2, \n",
    "             y=(base_count_prox_c_h115_1_bins[1]/(base_count_prox_c_h115_1_bins[2] + base_count_prox_c_h115_1_bins[1])) + \n",
    "               (base_count_prox_c_h115_2_bins[1]/(base_count_prox_c_h115_2_bins[2] + base_count_prox_c_h115_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'steelblue', ax = axs[0], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_c_h117_1_bins[0] + base_count_prox_c_h117_2_bins[0])/2, \n",
    "             y=(base_count_prox_c_h117_1_bins[1]/(base_count_prox_c_h117_1_bins[2] + base_count_prox_c_h117_1_bins[1])) + \n",
    "               (base_count_prox_c_h117_2_bins[1]/(base_count_prox_c_h117_2_bins[2] + base_count_prox_c_h117_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'orange', ax = axs[0], linewidth = 5)\n",
    "\n",
    "\n",
    "ax2[0] = sns.lineplot(x = (base_count_prox_c_h115_1_bins[0] + base_count_prox_c_h115_2_bins[0])/2, \n",
    "                   y = -np.log10(stat_vals_3), \n",
    "                   alpha=.8, color = 'tomato', ax = ax2[0], linewidth = 5)\n",
    "\n",
    "ax2[0].set_ylabel(\"-log10(p)\")\n",
    "ax2[0].axhline(-np.log10(0.05), linestyle = '--', color = 'black')\n",
    "\n",
    "axs[0].set_xlim(10, 390)\n",
    "ax2[0].set_xlim(10, 390)\n",
    "ax2[0].set(xticklabels=[])  # remove the tick labels\n",
    "ax2[0].tick_params(bottom=False)  # remove the ticks\n",
    "\n",
    "axs[0].set_ylim(0, 0.1)\n",
    "ax2[0].set_ylim(0,3)\n",
    "\n",
    "axs[0].set_ylabel(\"percent\")\n",
    "axs[0].set_xlabel(\"distance from cleavage site\")\n",
    "\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.savefig(figpath + 'nonA_C.png', dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare G usage - stats for 4SA vs 4SD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run t-test between percent nt ussage at each postion\n",
    "stat_vals_5 = stats.ttest_ind(\n",
    "                            [base_count_dist_g_h115_1_bins[1]/(base_count_dist_g_h115_1_bins[2] + base_count_dist_g_h115_1_bins[1]), base_count_dist_g_h115_2_bins[1]/(base_count_dist_g_h115_2_bins[2] + base_count_dist_g_h115_2_bins[1])],\n",
    "                            [base_count_dist_g_h117_1_bins[1]/(base_count_dist_g_h117_1_bins[2] + base_count_dist_g_h117_1_bins[1]), base_count_dist_g_h117_2_bins[1]/(base_count_dist_g_h117_2_bins[2] + (base_count_dist_g_h117_2_bins[1]))])[-1]\n",
    "\n",
    "#calculate FRD\n",
    "stat_vals_corr_5 = stats.false_discovery_control(stat_vals_5)\n",
    "\n",
    "#Run t-test between percent nt ussage at each postion\n",
    "stat_vals_3 = stats.ttest_ind(\n",
    "                            [base_count_prox_g_h115_1_bins[1]/(base_count_prox_g_h115_1_bins[2] + base_count_prox_g_h115_1_bins[1]), base_count_prox_g_h115_2_bins[1]/(base_count_prox_g_h115_2_bins[2] + base_count_prox_g_h115_2_bins[1])],\n",
    "                            [base_count_prox_g_h117_1_bins[1]/(base_count_prox_g_h117_1_bins[2] + base_count_prox_g_h117_1_bins[1]), base_count_prox_g_h117_2_bins[1]/(base_count_prox_g_h117_2_bins[2] + (base_count_prox_g_h117_2_bins[1]))])[-1]\n",
    "\n",
    "#calculate FRD\n",
    "stat_vals_corr_3 = stats.false_discovery_control(stat_vals_3)\n",
    "\n",
    "#############################################################\n",
    "\n",
    "#Figure for percent nt - 5' end A\n",
    "#Generate plot\n",
    "fig, (ax2, axs) = plt.subplots(2, 2, gridspec_kw={'height_ratios': [1, 3]}, figsize=(10, 5))\n",
    "\n",
    "#distal\n",
    "sns.lineplot(x = (base_count_dist_g_h112_1_bins[0] + base_count_dist_g_h112_2_bins[0])/2, \n",
    "             y=(base_count_dist_g_h112_1_bins[1]/(base_count_dist_g_h112_1_bins[2] + base_count_dist_g_h112_1_bins[1])) + \n",
    "               (base_count_dist_g_h112_2_bins[1]/(base_count_dist_g_h112_2_bins[2] + base_count_dist_g_h112_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'darkgreen', ax = axs[1], linewidth = 5)\n",
    "\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_g_h113_1_bins[0] + base_count_dist_g_h113_2_bins[0])/2, \n",
    "             y=(base_count_dist_g_h113_1_bins[1]/(base_count_dist_g_h113_1_bins[2] + base_count_dist_g_h113_1_bins[1])) + \n",
    "               (base_count_dist_g_h113_2_bins[1]/(base_count_dist_g_h113_2_bins[2] + base_count_dist_g_h113_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'tomato', ax = axs[1], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_g_h115_1_bins[0] + base_count_dist_g_h115_2_bins[0])/2, \n",
    "             y=(base_count_dist_g_h115_1_bins[1]/(base_count_dist_g_h115_1_bins[2] + base_count_dist_g_h115_1_bins[1])) + \n",
    "               (base_count_dist_g_h115_2_bins[1]/(base_count_dist_g_h115_2_bins[2] + base_count_dist_g_h115_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'steelblue', ax = axs[1], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_g_h117_1_bins[0] + base_count_dist_g_h117_2_bins[0])/2, \n",
    "             y=(base_count_dist_g_h117_1_bins[1]/(base_count_dist_g_h117_1_bins[2] + base_count_dist_g_h117_1_bins[1])) + \n",
    "               (base_count_dist_g_h117_2_bins[1]/(base_count_dist_g_h117_2_bins[2] + base_count_dist_g_h117_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'orange', ax = axs[1], linewidth = 5)\n",
    "\n",
    "\n",
    "ax2[1] = sns.lineplot(x = (base_count_dist_g_h115_1_bins[0] + base_count_dist_g_h115_2_bins[0])/2, \n",
    "                   y = -np.log10(stat_vals_5), \n",
    "                   alpha=.8, color = 'tomato', ax = ax2[1], linewidth = 5)\n",
    "\n",
    "ax2[1].set_ylabel(\"-log10(p)\")\n",
    "ax2[1].axhline(-np.log10(0.05), linestyle = '--', color = 'black')\n",
    "\n",
    "axs[1].set_xlim(10, 390)\n",
    "ax2[1].set_xlim(10, 390)\n",
    "ax2[1].set(xticklabels=[])  # remove the tick labels\n",
    "ax2[1].tick_params(bottom=False)  # remove the ticks\n",
    "\n",
    "axs[1].set_ylim(0, 0.1)\n",
    "ax2[1].set_ylim(0,3)\n",
    "\n",
    "axs[1].set_ylabel(\"percent\")\n",
    "axs[1].set_xlabel(\"distance from 3' end (nt)\")\n",
    "\n",
    "###############################################\n",
    "#proximal\n",
    "sns.lineplot(x = (base_count_prox_g_h112_1_bins[0] + base_count_prox_g_h112_2_bins[0])/2, \n",
    "             y=(base_count_prox_g_h112_1_bins[1]/(base_count_prox_g_h112_1_bins[2] + base_count_prox_g_h112_1_bins[1])) + \n",
    "               (base_count_prox_g_h112_2_bins[1]/(base_count_prox_g_h112_2_bins[2] + base_count_prox_g_h112_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'darkgreen', ax = axs[0], linewidth = 5)\n",
    "\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_g_h113_1_bins[0] + base_count_prox_g_h113_2_bins[0])/2, \n",
    "             y=(base_count_prox_g_h113_1_bins[1]/(base_count_prox_g_h113_1_bins[2] + base_count_prox_g_h113_1_bins[1])) + \n",
    "               (base_count_prox_g_h113_2_bins[1]/(base_count_prox_g_h113_2_bins[2] + base_count_prox_g_h113_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'tomato', ax = axs[0], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_g_h115_1_bins[0] + base_count_prox_g_h115_2_bins[0])/2, \n",
    "             y=(base_count_prox_g_h115_1_bins[1]/(base_count_prox_g_h115_1_bins[2] + base_count_prox_g_h115_1_bins[1])) + \n",
    "               (base_count_prox_g_h115_2_bins[1]/(base_count_prox_g_h115_2_bins[2] + base_count_prox_g_h115_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'steelblue', ax = axs[0], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_g_h117_1_bins[0] + base_count_prox_g_h117_2_bins[0])/2, \n",
    "             y=(base_count_prox_g_h117_1_bins[1]/(base_count_prox_g_h117_1_bins[2] + base_count_prox_g_h117_1_bins[1])) + \n",
    "               (base_count_prox_g_h117_2_bins[1]/(base_count_prox_g_h117_2_bins[2] + base_count_prox_g_h117_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'orange', ax = axs[0], linewidth = 5)\n",
    "\n",
    "\n",
    "ax2[0] = sns.lineplot(x = (base_count_prox_g_h115_1_bins[0] + base_count_prox_g_h115_2_bins[0])/2, \n",
    "                   y = -np.log10(stat_vals_3), \n",
    "                   alpha=.8, color = 'tomato', ax = ax2[0], linewidth = 5)\n",
    "\n",
    "ax2[0].set_ylabel(\"-log10(p)\")\n",
    "ax2[0].axhline(-np.log10(0.05), linestyle = '--', color = 'black')\n",
    "\n",
    "axs[0].set_xlim(10, 390)\n",
    "ax2[0].set_xlim(10, 390)\n",
    "ax2[0].set(xticklabels=[])  # remove the tick labels\n",
    "ax2[0].tick_params(bottom=False)  # remove the ticks\n",
    "\n",
    "axs[0].set_ylim(0, 0.1)\n",
    "ax2[0].set_ylim(0,3)\n",
    "\n",
    "axs[0].set_ylabel(\"percent\")\n",
    "axs[0].set_xlabel(\"distance from cleavage site\")\n",
    "\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.savefig(figpath + 'NonA_G.png', dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare U usage - stats for 4SA vs 4SD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run t-test between percent nt ussage at each postion\n",
    "stat_vals_5 = stats.ttest_ind(\n",
    "                            [base_count_dist_u_h115_1_bins[1]/(base_count_dist_u_h115_1_bins[2] + base_count_dist_u_h115_1_bins[1]), base_count_dist_u_h115_2_bins[1]/(base_count_dist_u_h115_2_bins[2] + base_count_dist_u_h115_2_bins[1])],\n",
    "                            [base_count_dist_u_h117_1_bins[1]/(base_count_dist_u_h117_1_bins[2] + base_count_dist_u_h117_1_bins[1]), base_count_dist_u_h117_2_bins[1]/(base_count_dist_u_h117_2_bins[2] + (base_count_dist_u_h117_2_bins[1]))])[-1]\n",
    "\n",
    "#calculate FRD\n",
    "stat_vals_corr_5 = stats.false_discovery_control(stat_vals_5)\n",
    "\n",
    "#Run t-test between percent nt ussage at each postion\n",
    "stat_vals_3 = stats.ttest_ind(\n",
    "                            [base_count_prox_u_h115_1_bins[1]/(base_count_prox_u_h115_1_bins[2] + base_count_prox_u_h115_1_bins[1]), base_count_prox_u_h115_2_bins[1]/(base_count_prox_u_h115_2_bins[2] + base_count_prox_u_h115_2_bins[1])],\n",
    "                            [base_count_prox_u_h117_1_bins[1]/(base_count_prox_u_h117_1_bins[2] + base_count_prox_u_h117_1_bins[1]), base_count_prox_u_h117_2_bins[1]/(base_count_prox_u_h117_2_bins[2] + (base_count_prox_u_h117_2_bins[1]))])[-1]\n",
    "\n",
    "#calculate FRD\n",
    "stat_vals_corr_3 = stats.false_discovery_control(stat_vals_3)\n",
    "\n",
    "#############################################################\n",
    "\n",
    "#Figure for percent nt - 5' end A\n",
    "#Generate plot\n",
    "fig, (ax2, axs) = plt.subplots(2, 2, gridspec_kw={'height_ratios': [1, 3]}, figsize=(10, 5))\n",
    "\n",
    "#distal\n",
    "sns.lineplot(x = (base_count_dist_u_h112_1_bins[0] + base_count_dist_u_h112_2_bins[0])/2, \n",
    "             y=(base_count_dist_u_h112_1_bins[1]/(base_count_dist_u_h112_1_bins[2] + base_count_dist_u_h112_1_bins[1])) + \n",
    "               (base_count_dist_u_h112_2_bins[1]/(base_count_dist_u_h112_2_bins[2] + base_count_dist_u_h112_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'darkgreen', ax = axs[1], linewidth = 5)\n",
    "\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_u_h113_1_bins[0] + base_count_dist_u_h113_2_bins[0])/2, \n",
    "             y=(base_count_dist_u_h113_1_bins[1]/(base_count_dist_u_h113_1_bins[2] + base_count_dist_u_h113_1_bins[1])) + \n",
    "               (base_count_dist_u_h113_2_bins[1]/(base_count_dist_u_h113_2_bins[2] + base_count_dist_u_h113_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'tomato', ax = axs[1], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_u_h115_1_bins[0] + base_count_dist_u_h115_2_bins[0])/2, \n",
    "             y=(base_count_dist_u_h115_1_bins[1]/(base_count_dist_u_h115_1_bins[2] + base_count_dist_u_h115_1_bins[1])) + \n",
    "               (base_count_dist_u_h115_2_bins[1]/(base_count_dist_u_h115_2_bins[2] + base_count_dist_u_h115_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'steelblue', ax = axs[1], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_dist_u_h117_1_bins[0] + base_count_dist_u_h117_2_bins[0])/2, \n",
    "             y=(base_count_dist_u_h117_1_bins[1]/(base_count_dist_u_h117_1_bins[2] + base_count_dist_u_h117_1_bins[1])) + \n",
    "               (base_count_dist_u_h117_2_bins[1]/(base_count_dist_u_h117_2_bins[2] + base_count_dist_u_h117_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'orange', ax = axs[1], linewidth = 5)\n",
    "\n",
    "\n",
    "ax2[1] = sns.lineplot(x = (base_count_dist_u_h115_1_bins[0] + base_count_dist_u_h115_2_bins[0])/2, \n",
    "                   y = -np.log10(stat_vals_5), \n",
    "                   alpha=.8, color = 'tomato', ax = ax2[1], linewidth = 5)\n",
    "\n",
    "ax2[1].set_ylabel(\"-log10(p)\")\n",
    "ax2[1].axhline(-np.log10(0.05), linestyle = '--', color = 'black')\n",
    "\n",
    "axs[1].set_xlim(10, 390)\n",
    "ax2[1].set_xlim(10, 390)\n",
    "ax2[1].set(xticklabels=[])  # remove the tick labels\n",
    "ax2[1].tick_params(bottom=False)  # remove the ticks\n",
    "\n",
    "axs[1].set_ylim(0, 0.1)\n",
    "ax2[1].set_ylim(0,3)\n",
    "\n",
    "axs[1].set_ylabel(\"percent\")\n",
    "axs[1].set_xlabel(\"distance from 3' end (nt)\")\n",
    "\n",
    "###############################################\n",
    "#proximal\n",
    "sns.lineplot(x = (base_count_prox_u_h112_1_bins[0] + base_count_prox_u_h112_2_bins[0])/2, \n",
    "             y=(base_count_prox_u_h112_1_bins[1]/(base_count_prox_u_h112_1_bins[2] + base_count_prox_u_h112_1_bins[1])) + \n",
    "               (base_count_prox_u_h112_2_bins[1]/(base_count_prox_u_h112_2_bins[2] + base_count_prox_u_h112_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'darkgreen', ax = axs[0], linewidth = 5)\n",
    "\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_u_h113_1_bins[0] + base_count_prox_u_h113_2_bins[0])/2, \n",
    "             y=(base_count_prox_u_h113_1_bins[1]/(base_count_prox_u_h113_1_bins[2] + base_count_prox_u_h113_1_bins[1])) + \n",
    "               (base_count_prox_u_h113_2_bins[1]/(base_count_prox_u_h113_2_bins[2] + base_count_prox_u_h113_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'tomato', ax = axs[0], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_u_h115_1_bins[0] + base_count_prox_u_h115_2_bins[0])/2, \n",
    "             y=(base_count_prox_u_h115_1_bins[1]/(base_count_prox_u_h115_1_bins[2] + base_count_prox_u_h115_1_bins[1])) + \n",
    "               (base_count_prox_u_h115_2_bins[1]/(base_count_prox_u_h115_2_bins[2] + base_count_prox_u_h115_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'steelblue', ax = axs[0], linewidth = 5)\n",
    "\n",
    "sns.lineplot(x = (base_count_prox_u_h117_1_bins[0] + base_count_prox_u_h117_2_bins[0])/2, \n",
    "             y=(base_count_prox_u_h117_1_bins[1]/(base_count_prox_u_h117_1_bins[2] + base_count_prox_u_h117_1_bins[1])) + \n",
    "               (base_count_prox_u_h117_2_bins[1]/(base_count_prox_u_h117_2_bins[2] + base_count_prox_u_h117_2_bins[1]))/2, \n",
    "                alpha=.6, color = 'orange', ax = axs[0], linewidth = 5)\n",
    "\n",
    "\n",
    "ax2[0] = sns.lineplot(x = (base_count_prox_u_h115_1_bins[0] + base_count_prox_u_h115_2_bins[0])/2, \n",
    "                   y = -np.log10(stat_vals_3), \n",
    "                   alpha=.8, color = 'tomato', ax = ax2[0], linewidth = 5)\n",
    "\n",
    "ax2[0].set_ylabel(\"-log10(p)\")\n",
    "ax2[0].axhline(-np.log10(0.05), linestyle = '--', color = 'black')\n",
    "\n",
    "axs[0].set_xlim(10, 390)\n",
    "ax2[0].set_xlim(10, 390)\n",
    "ax2[0].set(xticklabels=[])  # remove the tick labels\n",
    "ax2[0].tick_params(bottom=False)  # remove the ticks\n",
    "\n",
    "axs[0].set_ylim(0, 0.1)\n",
    "ax2[0].set_ylim(0,3)\n",
    "\n",
    "axs[0].set_ylabel(\"percent\")\n",
    "axs[0].set_xlabel(\"distance from cleavage site\")\n",
    "\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.savefig(figpath + 'nonA_U.png', dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I need a function to determine which reads have nonA usage either at distal end or in general"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
